astropy/coordinates/sky_coordinate.py

```diff
@@ -1205,46 +1213,80 @@ def _parse_coordinate_arg(coords, frame, units):
             if isinstance(coord, six.string_types):
                 coord1 = coord.split()
                 if len(coord1) == 6:
                     coord1 = (' '.join(coord1[:3]), ' '.join(coord1[3:]))
                 coord = coord1
-
             vals.append(coord)  # This assumes coord is a sequence at this point
 
-        # Do some basic validation of the list elements: all have a length and all
-        # lengths the same
-        try:
-            n_coords = sorted(set(len(x) for x in vals))
-        except:
-            raise ValueError('One or more elements of input sequence does not have a length')
-
-        if len(n_coords) > 1:
-            raise ValueError('Input coordinate values must have same number of elements, found {0}'
-                             .format(n_coords))
-        n_coords = n_coords[0]
-
-        # Must have no more coord inputs than representation attributes
-        if n_coords > n_attr_names:
-            raise ValueError('Input coordinates have {0} values but {1} representation '
-                             'only accepts {2}'
-                             .format(n_coords, frame.representation.get_name(), n_attr_names))
-
-        # Now transpose vals to get [(v1_0 .. v1_N), (v2_0 .. v2_N), (v3_0 .. v3_N)]
-        # (ok since we know it is exactly rectangular).  (Note: can't just use zip(*values)
-        # because Longitude et al distinguishes list from tuple so [a1, a2, ..] is needed
-        # while (a1, a2, ..) doesn't work.
-        values = [list(x) for x in zip(*vals)]
-
-        if is_scalar:
-            values = [x[0] for x in values]
-
+        if hascoordobjs:
+            # this parsing path is used when there are coordinate-like objects
+            # in the list - instead of creating lists of values, we create
+            # SkyCoords from the list elements and then combine them.
+            scs = [SkyCoord(coord, **init_kwargs) for coord in coords]
+
+            # now check that they're all self-consistent in their frame attributes
+            # and frame name
+            if len(set((sc.frame.name for sc in scs if hasattr(sc, 'frame')))) > 1:
+                frames = [sc.frame.name if hasattr(sc, 'frame') else None for sc in scs]
+                raise ValueError("List of inputs have different frames: {0}".format(frames))
+            for fattrnm in FRAME_ATTR_NAMES_SET():
+                vals = [getattr(sc, fattrnm) for sc in scs]
+                for val in vals[1:]:
+                    if val != vals[0]:
+                        raise ValueError("List of inputs don't give consistent "
+                                         "frame attribute {0}: {1}".format(fattrnm, vals))
+
+                valid_kwargs[fattrnm] = getattr(scs[0], fattrnm)
+
+            # Now combine the values, to be used below
+            values = []
+            for aname in frame_attr_names:
+                values.append([])
```

Comment:
This will end up generating individual `Quantity` objects for every element of each `sc`, so if could really blow up for large input `SkyCoord` objects.  Here's an idea, completely untested:
```
for aname in frame_attr_names:
    attr_vals = []
    for sc in scs:
        attr_val = getattr(sc, aname)
        attr_vals.append(attr_val.reshape(1,) if sc.isscalar else attr_val)
    concat_vals = np.concatenate(attr_vals)
    # Hack because np.concatenate doesn't fully work with Quantity
    if isinstance(concat_vals, Quantity):
        concat_vals._unit = attr_val.unit
    values.append(concat_vals)
```